lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,259 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/generated/resolver.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isLoopStatement(AstNode node) => node is DoStatement ||
+      node is ForEachStatement ||
+      node is ForStatement ||
+      node is WhileStatement;
+
+_needsLocalEnvironment(AstNode node) =>
+    node is ClassDeclaration ||
+    node is CompilationUnit ||
+    node is ConstructorDeclaration ||
+    node is FunctionDeclaration ||
+    node is MethodDeclaration;
+
+abstract class ConditionScopeVisitor extends UnifyingAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
+    node.visitChildren(this);
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    visitCondition(node.condition);
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: (node) => _isLoopStatement(node))
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.variables?.accept(this);
+    node.initialization?.accept(this);
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _removeLocalEnvironment();
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
```

Comment:
It works well, if you have a break, what you know is the j < 10 is not necessary false after the ForStatement, but something that we should add is when you find a label, for example outer:, we should undefine all